File: apps/mobile/android/app/src/main/java/br/com/aipinho/mobile/network/GovernanceClient.kt
package br.com.aipinho.mobile.network

import br.com.aipinho.mobile.models.ConnectionProfile

class GovernanceClient(profile: ConnectionProfile, token: () -> String?) : BaseApiClient(profile, profile.corePort, token, timeoutMs = 30_000) {
    fun health() = get("/api/v1/config/health")
    fun effectivePolicy() = get("/api/v1/config/effective-policy")
    fun workspaces() = get("/api/v1/config/workspaces")
    fun permissionMatrix() = get("/api/v1/config/permission-matrix")
    fun changes() = get("/api/v1/config/changes")
    fun backups() = get("/api/v1/config/backups")
    fun flowRules() = get("/api/v1/workspace-flows/rules")
    fun createWorkspace(workspaceId: String, label: String, rootPath: String, role: String, permissionsJson: String) = post(
        "/api/v1/config/workspaces",
        """
        {
          "workspace_id":"${escapeJson(workspaceId)}",
          "root_path":"${escapeJson(rootPath)}",
          "role":"${escapeJson(role)}",
          "human_label":"${escapeJson(label)}",
          "reason":"mobile_governance_console_workspace_change",
          "approval_required":true,
          "enabled":true,
          "permissions":${permissionsJson.ifBlank { "{}" }},
          "evidence":["mobile_governance_console"]
        }
        """.trimIndent()
    )
    fun approveChange(changeId: String) = post("/api/v1/config/changes/${escapePath(changeId)}/approve")
    fun applyChange(changeId: String) = post("/api/v1/config/changes/${escapePath(changeId)}/apply")
    fun rollback(backupId: String) = post("/api/v1/config/rollback/${escapePath(backupId)}")
    fun flowPlan(operation: String, sourceWorkspaceId: String, targetWorkspaceId: String) = post(
        "/api/v1/workspace-flows/plan",
        """
        {
          "operation":"${escapeJson(operation)}",
          "source":{"workspace_id":"${escapeJson(sourceWorkspaceId)}","path":""},
          "target":{"workspace_id":"${escapeJson(targetWorkspaceId)}","path":""},
          "requested_by":{"type":"user","id":"mobile"}
        }
        """.trimIndent()
    )

    private fun escapeJson(value: String): String =
        value.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n").replace("\r", "\\r")

    private fun escapePath(value: String): String = java.net.URLEncoder.encode(value, Charsets.UTF_8.name())
}

File: apps/launcher/ui/api/governance_client.py
from __future__ import annotations

from apps.launcher.ui.api.base_client import ApiResult, BaseClient


class GovernanceClient(BaseClient):
    def health(self) -> ApiResult:
        return self.get("/api/v1/config/health")

    def effective_policy(self) -> ApiResult:
        return self.get("/api/v1/config/effective-policy")

    def workspaces(self) -> ApiResult:
        return self.get("/api/v1/config/workspaces")

    def workspace_roles(self) -> ApiResult:
        return self.get("/api/v1/config/workspace-roles")

    def permission_matrix(self) -> ApiResult:
        return self.get("/api/v1/config/permission-matrix")

    def providers(self) -> ApiResult:
        return self.get("/api/v1/config/providers")

    def agents(self) -> ApiResult:
        return self.get("/api/v1/config/agents")

    def changes(self) -> ApiResult:
        return self.get("/api/v1/config/changes")

    def backups(self) -> ApiResult:
        return self.get("/api/v1/config/backups")

    def preview_workspace(self, workspace_id: str, path: str, role: str) -> ApiResult:
        return self.post(
            "/api/v1/config/workspaces/preview",
            {"workspace_id": workspace_id, "path": path, "role": role},
        )

    def create_workspace(
        self,
        *,
        workspace_id: str,
        root_path: str,
        role: str,
        human_label: str = "",
        permissions: dict[str, str] | None = None,
    ) -> ApiResult:
        return self.post(
            "/api/v1/config/workspaces",
            {
                "workspace_id": workspace_id,
                "root_path": root_path,
                "role": role,
                "human_label": human_label,
                "reason": "launcher_governance_console_workspace_change",
                "approval_required": True,
                "enabled": True,
                "permissions": permissions or {},
                "evidence": ["launcher_governance_console"],
            },
        )

    def approve_change(self, change_id: str) -> ApiResult:
        return self.post(f"/api/v1/config/changes/{change_id}/approve")

    def apply_change(self, change_id: str) -> ApiResult:
        return self.post(f"/api/v1/config/changes/{change_id}/apply")

    def cancel_change(self, change_id: str) -> ApiResult:
        return self.post(f"/api/v1/config/changes/{change_id}/cancel")

    def rollback(self, backup_id: str) -> ApiResult:
        return self.post(f"/api/v1/config/rollback/{backup_id}")

    def flow_rules(self) -> ApiResult:
        return self.get("/api/v1/workspace-flows/rules")

    def flow_plan(self, payload: dict[str, object]) -> ApiResult:
        return self.post("/api/v1/workspace-flows/plan", payload)

    def flow_plan_detail(self, flow_plan_id: str) -> ApiResult:
        return self.get(f"/api/v1/workspace-flows/plans/{flow_plan_id}")

    def approve_flow_plan(self, flow_plan_id: str) -> ApiResult:
        return self.post(f"/api/v1/workspace-flows/plans/{flow_plan_id}/approve")

    def deny_flow_plan(self, flow_plan_id: str) -> ApiResult:
        return self.post(f"/api/v1/workspace-flows/plans/{flow_plan_id}/deny")

    def execute_flow_plan(self, flow_plan_id: str) -> ApiResult:
        return self.post(f"/api/v1/workspace-flows/plans/{flow_plan_id}/execute")
